### ISSUE TITLE
Improve shutdown sequence of DebugAdapter

### ISSUE THREAD

**Issue by daviwil** [PowerShell/PowerShellEditorServices#110]
There is a timing issue with the shutdown sequence of the DebugAdapter which occasionally causes test hangs because the `terminated` event is not received.  My guess is that this event never gets fully sent because of the timing of the shutdown sequence relative to some event handlers that get fired at that time.  Compare the following two log files, the invocation state changes are interesting.

[GoodRun.log.txt](https://github.com/PowerShell/PowerShellEditorServices/files/85231/GoodRun.log.txt)
[BadRun.log.txt](https://github.com/PowerShell/PowerShellEditorServices/files/85232/BadRun.log.txt)

There are some additional notes in the BadRun.log.txt file at the end.

It's likely that the shutdown sequence will need to be rethought so that the order of events is predictable and all messages get sent before the process terminates.

**Comment by rkeithhill** [comment]
BTW in StdioClientChannel.cs I wonder if we should change wrap this call:
```csharp
            this.serviceProcess.Kill();
```
in a try/catch.  I've observed this line throwing at times when running the unit tests (usually when I've screwed up something) because the service has already exited and then we try to kill it.

**Comment by daviwil** [comment]
Definitely, good point. I'll take that into account when I work on this. I've got some stuff I've already started to improve the async model so that CancellationTokens and failures flow through more easily (also with a fix for the CPU spin in MesaageReader). Will try to get that in for 0.5.0.

**Comment by rkeithhill** [comment]
I also notice this issue when I'm manually stepping through a unit test.  By the time, I get to the line to send the DisconnectRequest, the process has already terminated which throws and causes the unit test to fail.  These usually don't fail this when when not run under the debugger.

**Comment by daviwil** [comment]
Yeah, it needs to work correctly for sure.  I keep seeing an annoying ThreadAbortException when running the tests in the debugger, is that what you're seeing too?

**Comment by rkeithhill** [comment]
I don't think that was what I was seeing but I was "experimenting" and likely causing the host to exit prematurely.  :-)

**Comment by andyleejordan** [comment]
Believed resolved — DAP shutdown is handled by `DisconnectHandler`. Closing as completed; reopen if shutdown still hangs.

---
_Drafted by GitHub Copilot (Claude Opus 4.8) at @andyleejordan's request._
